File: src/types/stock.ts
export interface StockQuote {
  symbol: string;
  longName: string;
  price: number;
  previousClose: number;
  change: number;
  changePercent: number;
  volume: number;
  marketCap: number;
  fiftyTwoWeekHigh: number;
  fiftyTwoWeekLow: number;
  currency: string;
  marketState: string;
  meta?: import("./data-meta").DataMeta;
}

export interface FinancialMetrics {
  trailingPE: number | null;
  forwardPE: number | null;
  pegRatio: number | null;
  priceToBook: number | null;
  dividendYield: number | null;
  grossMargins: number | null;
  operatingMargins: number | null;
  profitMargins: number | null;
  revenueGrowth: number | null;
  earningsGrowth: number | null;
  returnOnEquity: number | null;
  debtToEquity: number | null;
  freeCashflow: number | null;
  marketCap: number | null;
  enterpriseValue: number | null;
  beta: number | null;
  floatShares?: number | null;
  sharesOutstanding?: number | null;
  sharesShort?: number | null;
  sharesShortPriorMonth?: number | null;
  shortRatio?: number | null;
  shortPercentOfFloat?: number | null;
  dateShortInterest?: number | null;
  meta?: import("./data-meta").DataMeta;
}

export interface EarningsRecord {
  date: string;
  epsActual: number | null;
  epsEstimate: number | null;
  surprise: number | null;
  surprisePercent: number | null;
  revenue: number | null;
  revenueEstimate: number | null;
}

export interface ChartDataPoint {
  time: number;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
}

export interface StockSearchResult {
  symbol: string;
  name: string;
  exchange: string;
  type: string;
}

File: python-api/routers/stock.py
from fastapi import APIRouter, HTTPException
from datetime import datetime
import yfinance as yf
from utils import safe_num

router = APIRouter()


@router.get("/{symbol}/financials")
def get_financials(symbol: str):
    try:
        t = yf.Ticker(symbol)
        info = t.info
    except Exception as e:
        raise HTTPException(status_code=502, detail=str(e))

    return {
        "symbol": symbol.upper(),
        "longName": info.get("longName", symbol.upper()),
        "price": safe_num(info.get("currentPrice") or info.get("regularMarketPrice")),
        "previousClose": safe_num(info.get("previousClose") or info.get("regularMarketPreviousClose")),
        "marketCap": safe_num(info.get("marketCap")),
        "trailingPE": safe_num(info.get("trailingPE")),
        "forwardPE": safe_num(info.get("forwardPE")),
        "pegRatio": safe_num(info.get("pegRatio")),
        "priceToBook": safe_num(info.get("priceToBook")),
        "dividendYield": safe_num(info.get("dividendYield")),
        "grossMargins": safe_num(info.get("grossMargins")),
        "operatingMargins": safe_num(info.get("operatingMargins")),
        "profitMargins": safe_num(info.get("profitMargins")),
        "revenueGrowth": safe_num(info.get("revenueGrowth")),
        "earningsGrowth": safe_num(info.get("earningsGrowth")),
        "returnOnEquity": safe_num(info.get("returnOnEquity")),
        "debtToEquity": safe_num(info.get("debtToEquity")),
        "freeCashflow": safe_num(info.get("freeCashflow")),
        "enterpriseValue": safe_num(info.get("enterpriseValue")),
        "beta": safe_num(info.get("beta")),
        "fiftyTwoWeekHigh": safe_num(info.get("fiftyTwoWeekHigh")),
        "fiftyTwoWeekLow": safe_num(info.get("fiftyTwoWeekLow")),
        "volume": safe_num(info.get("volume")),
        "floatShares": safe_num(info.get("floatShares")),
        "sharesOutstanding": safe_num(info.get("sharesOutstanding")),
        "sharesShort": safe_num(info.get("sharesShort")),
        "sharesShortPriorMonth": safe_num(info.get("sharesShortPriorMonth")),
        "shortRatio": safe_num(info.get("shortRatio")),
        "shortPercentOfFloat": safe_num(info.get("shortPercentOfFloat")),
        "dateShortInterest": safe_num(info.get("dateShortInterest")),
        "sector": info.get("sector"),
        "industry": info.get("industry"),
        "meta": {
            "source": "yfinance (Yahoo Finance)",
            "asOf": datetime.utcnow().date().isoformat(),
            "fetchedAt": datetime.utcnow().isoformat() + "Z",
            "frequency": "on-demand",
            "confidence": "high",
        },
    }


@router.get("/{symbol}/earnings")
def get_earnings(symbol: str):
    try:
        t = yf.Ticker(symbol)
        hist = t.earnings_history
    except Exception as e:
        raise HTTPException(status_code=502, detail=str(e))

    if hist is None or hist.empty:
        return {"symbol": symbol.upper(), "earnings": []}

    records = []
    for _, row in hist.iterrows():
        records.append({
            "date": str(row.name) if hasattr(row, "name") else str(row.get("Earnings Date", "")),
            "epsActual": safe_num(row.get("Reported EPS")),
            "epsEstimate": safe_num(row.get("EPS Estimate")),
            "surprise": safe_num(row.get("Surprise(%)")),
        })

    return {"symbol": symbol.upper(), "earnings": records}


@router.get("/{symbol}/search")
def search_stock(symbol: str, q: str = ""):
    query = q or symbol
    try:
        results = yf.search(query, max_results=8)
    except Exception:
        results = {"quotes": []}

    quotes = results.get("quotes", [])
    return [
        {
            "symbol": item.get("symbol", ""),
            "name": item.get("longname") or item.get("shortname", ""),
            "exchange": item.get("exchange", ""),
            "type": item.get("quoteType", ""),
        }
        for item in quotes
        if item.get("symbol")
    ]
